server/handlers/authentication.go

```diff
@@ -1,158 +1,100 @@
-				http.Redirect(w, r, next, http.StatusSeeOther)
-				return
-			}
-		}
-
-		utils.ParseHTML(w, "assets-v1/templates/src/pages/oauth_portal/login.html",
-			map[string]interface{}{
-				"HasNext":  next != "",
-				"Next":     next,
-				"ProxyURL": os.Getenv("PROXY_URL"),
-			},
-		)
-		return
+		a.getLoginHandler(w, r)
 	case "POST":
-		next := r.URL.Query().Get("next")
-		username := r.FormValue("username")
-		password := r.FormValue("password")
-		// login the user
-		rUser, err := service.LoginUser(username, password)
-		if err != nil {
-			t, errT := template.ParseFiles("assets-v1/templates/src/pages/oauth_portal/login.html")
-			if errT != nil {
-				http.Error(w, errT.Error(), http.StatusInternalServerError)
-				return
-			}
-			t.Execute(w, map[string]interface{}{
-				"HasNext": next != "",
-				"Next":    next,
-				"Error":   err.Error(),
-			})
-			return
-		}
-		// set the token cookie
-		token, ok := rUser["token"].(string)
-		if !ok {
-			t, err := template.ParseFiles("assets-v1/templates/src/pages/oauth_portal/login.html")
-			if err != nil {
-				http.Error(w, err.Error(), http.StatusInternalServerError)
-				return
-			}
-			t.Execute(w, map[string]interface{}{
-				"HasNext": next != "",
-				"Next":    next,
-				"Error":   "could not get token",
-			})
-			return
-		}
-		http.SetCookie(w, &http.Cookie{
-			Name:  "token",
-			Value: token,
-			Path:  "/",
-		})
-		// redirect to next page - here the user already knows their pseudo profile
-		// when they registered
-		http.Redirect(w, r, next, http.StatusSeeOther)
-		return
+		a.postLoginHandler(w, r)
 	default:
 		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
-		return
 	}
 }
-func Register(w http.ResponseWriter, r *http.Request) {
-	service := r.Context().Value("OauthService").(*svc.Service)
-
-	switch r.Method {
-	case "GET":
-		next := r.URL.Query().Get("next")
-		if next == "" {
-			next = "/"
-		}
-		// check if the user is already logged in
-		token, err := r.Cookie("token")
-		if token != nil && err == nil {
-			user, err := service.GetUserByToken(token.Value)
-			if err == nil && user != nil {
-				http.Redirect(w, r, next, http.StatusSeeOther)
-				return
-			}
-		}
 
-		utils.ParseHTML(w, "registerClient.html",
-			map[string]interface{}{
-				"HasNext":  next != "",
-				"Next":     next,
-				"ProxyURL": os.Getenv("PROXY_URL"),
-			},
-		)
-		return
-	case "POST":
-		next := r.URL.Query().Get("next")
-		username := r.FormValue("username")
-		password := r.FormValue("password")
-		// login the user
-		rUser, err := service.LoginUser(username, password)
-		if err != nil {
-			utils.ParseHTML(w, "registerClient.html",
-				map[string]interface{}{
-					"HasNext":  next != "",
-					"Next":     next,
-					"Error":    err.Error(),
-					"ProxyURL": os.Getenv("PROXY_URL"),
-				},
-			)
+func (a *authenticationHandlerImpl) getLoginHandler(w http.ResponseWriter, r *http.Request) {
+	next := r.URL.Query().Get("next")
```

Comment:
Next is not a pointer, it can't have a nil value. However, if the value is a special character, it needs to be tested via an integration test because it relates to browser behavior handling redirection to special character URL